File: python_workflow/snakemake/scripts/old_unused_scripts/revise.R
#!/usr/bin/env Rscript

# Load required packages
library(cluster)

# Parse arguments from command line
args <- commandArgs(trailingOnly = TRUE)
input_file <- args[1]
output_file <- args[2]
sample <- args[3]

# Hierarchical clustering helper function
hc <- function(x, k, d.meth = "euclidean", ...) {
  list(cluster = cutree(hclust(dist(x, method = d.meth), ...), k = k))
}

# Function to calculate clonal and subclonal distances
clon_sub_distance <- function(df, by_state = TRUE) {
  if (by_state) {
    df$clonal_cn_state <- paste0(df$major_cn, '_', df$minor_cn)
    states <- unique(df$clonal_cn_state)
  } else {
    states <- 'all'
    df$clonal_cn_state <- "all"
  }

  df$dist_clon <- NA
  df$dist_sub <- NA
  df$mean_count <- NA
  df$sd_count <- NA
  df$mean_count_sub <- NA
  df$sd_count_sub <- NA

  for (state in states) {
    if (all(is.na(df$best_cn[df$categ == "clonal" & df$clonal_cn_state == state]))) {
      next  # Skip invalid states
    }
    
    min_clonal_cn <- min(df$best_cn[df$categ == "clonal" & df$clonal_cn_state == state], na.rm = TRUE)
    inds <- which(df$best_cn == min_clonal_cn & df$categ == "clonal" & df$clonal_cn_state == state)
    
    mean_count <- mean((1 / df$vaf[inds]) / df$total_cn[inds], na.rm = TRUE)
    sd_count <- sd((1 / df$vaf[inds]) / df$total_cn[inds], na.rm = TRUE)
    mean_count_sub <- mean((1 / df$vaf[df$clonal_cn_state == state & df$categ == "subclonal1"]) /
                             df$total_cn[df$clonal_cn_state == state & df$categ == "subclonal1"], na.rm = TRUE)
    sd_count_sub <- sd((1 / df$vaf[df$clonal_cn_state == state & df$categ == "subclonal1"]) /
                         df$total_cn[df$clonal_cn_state == state & df$categ == "subclonal1"], na.rm = TRUE)
    inds_state <- which(df$clonal_cn_state == state)
    if (length(inds_state) > 10) {
      df$mean_count[inds_state] <- mean_count
      df$sd_count[inds_state] <- sd_count
      df$mean_count_sub[inds_state] <- mean_count_sub
      df$sd_count_sub[inds_state] <- sd_count_sub

      df$dist_clon[inds_state] <- abs((1 / df$vaf[inds_state]) / df$total_cn[inds_state] - df$mean_count[inds_state]) / df$sd_count[inds_state]
      df$dist_sub[inds_state] <- abs((1 / df$vaf[inds_state]) / df$total_cn[inds_state] - df$mean_count_sub[inds_state]) / df$sd_count_sub[inds_state]

      df$dist_clon[inds_state[df$best_cn[inds_state] > min_clonal_cn]] <- 0
    }
  }
  return(df)
}

# Function to cluster subclones
cluster_subclones <- function(df, sample, output_dir, min_vaf_diff = 0.05) {
  # Extract values for clustering
  subclonal_rows <- grepl("subclonal", df$categ)
  vals <- (1 / df$vaf[subclonal_rows]) / df$total_cn[subclonal_rows]
  
  # Filter valid values
  valid_indices <- which(is.finite(vals) & vals > 0)
  vals <- vals[valid_indices]
  
  if (length(vals) == 0) {
    stop("No valid data for clustering. Check VAF and total CN values.")
  }

  # Prepare data for clustering
  df_x <- data.frame(x = log2(vals + 1))

  # Perform clustering using clusGap
  gs <- cluster::clusGap(df_x, FUNcluster = hc, K.max = 5, B = 20, verbose = FALSE)
  best_k <- cluster::maxSE(gs$Tab[, "gap"], gs$Tab[, "SE.sim"], method = "Tibs2001SEmax")
  hclust <- hclust(dist(df_x))
  cluster_inds <- cutree(hclust, k = best_k)

  # Calculate mean values per cluster
  mean_vec <- numeric()
  for (j in 1:best_k) {
    mean_vec <- c(mean_vec, mean(vals[cluster_inds == j], na.rm = TRUE))
  }

  # Create subclonal CCF dataframe
  df_subclonal_CCF <- data.frame(CCFs_subclone = 1 / mean_vec, index = 1:length(mean_vec))
  if (nrow(df_subclonal_CCF) > 1) {
    df_subclonal_CCF <- df_subclonal_CCF[order(-df_subclonal_CCF$CCFs_subclone), ]
  }
  df_subclonal_CCF$cluster <- 1:nrow(df_subclonal_CCF)
  write.table(df_subclonal_CCF, paste0(output_dir, '/', sample, '_CCF_subclones.txt'), row.names = FALSE, sep = '\t', quote = FALSE)

  # Update `df$categ` with subclone clusters
  indices <- df_subclonal_CCF$cluster[match(cluster_inds, df_subclonal_CCF$index)]
  replacement_indices <- which(subclonal_rows)[valid_indices]  # Match valid rows
  if (length(indices) != length(replacement_indices)) {
    stop("Replacement length mismatch: check indices and subclonal categories.")
  }
  df$categ[replacement_indices] <- paste0("subclonal", indices)
  return(df)
}

# Main revise function
revise <- function(sample, input_file, output_file, n_sub_thresh = 20, distance_scale = 3) {
  df <- read.delim(input_file)

  has_subclone <- FALSE
  if (sum(grepl('subclonal', df$categ), na.rm = TRUE) > n_sub_thresh) {
    print('Clustering subclones (Step 1)')
    df <- cluster_subclones(df, sample, output_dir = dirname(output_file))
    has_subclone <- TRUE
  } else {
    df$categ[grepl('subclonal', df$categ)] <- 'undefined'
  }

  if (!has_subclone) {
    write.table(df, file = output_file, row.names = FALSE, sep = '\t', quote = FALSE)
    return(0)
  }

  df <- clon_sub_distance(df)

  # Recategorize subclonal1 to clonal if closer
  inds <- which(df$dist_clon < distance_scale * df$dist_sub & df$categ == "subclonal1" & df$max_likelihood > 0.02)
  if (length(inds) > 0) {
    df$categ[inds] <- 'clonal'
    print('Re-clustering subclones (Step 2)')
    df <- cluster_subclones(df, sample, output_dir = dirname(output_file))
    df <- clon_sub_distance(df)
  }

  # Recategorize undefined to subclonal1 if closer
  inds <- which(df$dist_clon > distance_scale * df$dist_sub & !grepl('subclonal', df$categ) & df$dist_clon > 0 & df$vaf < 1 / df$total_cn)
  if (length(inds) > 0) {
    df$categ[inds] <- 'subclonal1'
    print('Re-clustering subclones (Step 3)')
    df <- cluster_subclones(df, sample, output_dir = dirname(output_file))
    df <- clon_sub_distance(df)
  }

  if (sum(grepl('subclonal', df$categ), na.rm = TRUE) > n_sub_thresh) {
    print('Final subclone clustering (Step 4)')
    df <- cluster_subclones(df, sample, output_dir = dirname(output_file))
  }

  write.table(df, file = output_file, row.names = FALSE, sep = '\t', quote = FALSE)
  return(0)
}

# Execute the revise function with parsed arguments
revise(sample, input_file, output_file)

File: python_workflow/snakemake/scripts/revise.py
#!/usr/bin/env python3

import pandas as pd
import numpy as np
from scipy.cluster.hierarchy import linkage, cut_tree, fcluster
from scipy.spatial.distance import pdist
from scipy.stats import betabinom
import argparse
import os

def hc(x, k, d_meth='euclidean', **kwargs):
    Z = linkage(x, method='ward', metric=d_meth)
    return {'cluster': cut_tree(Z, n_clusters=k).flatten()}

def clon_sub_distance(df, by_state=True):
    if by_state:
        df['clonal_cn_state'] = df['major_cn'].astype(str) + '_' + df['minor_cn'].astype(str)
        states = df['clonal_cn_state'].unique()
    else:
        states = ['all']
        df['clonal_cn_state'] = 'all'

    df['dist_clon'] = np.nan
    df['dist_sub'] = np.nan
    df['mean_count'] = np.nan
    df['sd_count'] = np.nan
    df['mean_count_sub'] = np.nan
    df['sd_count_sub'] = np.nan

    for state in states:
        state_df = df[(df['categ'] == 'clonal') & (df['clonal_cn_state'] == state)]
        if state_df['best_cn'].isna().all():
            continue

        min_clonal_cn = state_df['best_cn'].min()
        inds = df[(df['best_cn'] == min_clonal_cn) & (df['categ'] == 'clonal') & (df['clonal_cn_state'] == state)].index

        mean_count = ((1 / df.loc[inds, 'vaf']) / df.loc[inds, 'total_cn']).mean()
        sd_count = ((1 / df.loc[inds, 'vaf']) / df.loc[inds, 'total_cn']).std()
        
        # Fix: Separate filtering and column access
        subclonal1_mask = (df['clonal_cn_state'] == state) & (df['categ'] == 'subclonal1')
        mean_count_sub = ((1 / df.loc[subclonal1_mask, 'vaf']) / df.loc[subclonal1_mask, 'total_cn']).mean()
        sd_count_sub = ((1 / df.loc[subclonal1_mask, 'vaf']) / df.loc[subclonal1_mask, 'total_cn']).std()

        inds_state = df[df['clonal_cn_state'] == state].index
        if len(inds_state) > 10:
            df.loc[inds_state, 'mean_count'] = mean_count
            df.loc[inds_state, 'sd_count'] = sd_count
            df.loc[inds_state, 'mean_count_sub'] = mean_count_sub
            df.loc[inds_state, 'sd_count_sub'] = sd_count_sub

            df.loc[inds_state, 'dist_clon'] = np.abs((1 / df.loc[inds_state, 'vaf']) / df.loc[inds_state, 'total_cn'] - df.loc[inds_state, 'mean_count']) / df.loc[inds_state, 'sd_count']
            df.loc[inds_state, 'dist_sub'] = np.abs((1 / df.loc[inds_state, 'vaf']) / df.loc[inds_state, 'total_cn'] - df.loc[inds_state, 'mean_count_sub']) / df.loc[inds_state, 'sd_count_sub']

            df.loc[inds_state[df.loc[inds_state, 'best_cn'] > min_clonal_cn], 'dist_clon'] = 0

    return df

def cluster_subclones(df, sample, output_dir, min_vaf_diff=0.05):
    subclonal_rows = df['categ'].str.contains('subclonal')
    vals = (1 / df.loc[subclonal_rows, 'vaf']) / df.loc[subclonal_rows, 'total_cn']

    valid_indices = vals.index[np.isfinite(vals) & (vals > 0)]
    vals = vals[valid_indices]

    if len(vals) == 0:
        raise ValueError("No valid data for clustering. Check VAF and total CN values.")

    df_x = pd.DataFrame({'x': np.log2(vals + 1)})

    # Perform clustering using hierarchical clustering
    Z = linkage(df_x, method='ward')
    best_k = 5  # Placeholder for actual gap statistic calculation
    cluster_inds = fcluster(Z, best_k, criterion='maxclust')

    mean_vec = []
    for j in range(1, best_k + 1):
        mean_vec.append(vals[cluster_inds == j].mean())

    df_subclonal_CCF = pd.DataFrame({'CCFs_subclone': 1 / np.array(mean_vec), 'index': range(1, len(mean_vec) + 1)})
    if len(df_subclonal_CCF) > 1:
        df_subclonal_CCF = df_subclonal_CCF.sort_values(by='CCFs_subclone', ascending=False)
    df_subclonal_CCF['cluster'] = range(1, len(df_subclonal_CCF) + 1)
    df_subclonal_CCF.to_csv(os.path.join(output_dir, f'{sample}_CCF_subclones.txt'), sep='\t', index=False)

    indices = df_subclonal_CCF['cluster'].values[cluster_inds - 1]
    replacement_indices = subclonal_rows[valid_indices].index
    if len(indices) != len(replacement_indices):
        raise ValueError("Replacement length mismatch: check indices and subclonal categories.")
    df.loc[replacement_indices, 'categ'] = ['subclonal' + str(i) for i in indices]

    return df

def revise(sample, input_file, output_file, n_sub_thresh=20, distance_scale=3):
    df = pd.read_csv(input_file, sep='\t')

    has_subclone = False
    if df['categ'].str.contains('subclonal').sum() > n_sub_thresh:
        print('Clustering subclones (Step 1)')
        df = cluster_subclones(df, sample, output_dir=os.path.dirname(output_file))
        has_subclone = True
    else:
        df.loc[df['categ'].str.contains('subclonal'), 'categ'] = 'undefined'

    if not has_subclone:
        df.to_csv(output_file, sep='\t', index=False)
        return 0

    df = clon_sub_distance(df)

    inds = df[(df['dist_clon'] < distance_scale * df['dist_sub']) & (df['categ'] == "subclonal1") & (df['max_likelihood'] > 0.02)].index
    if len(inds) > 0:
        df.loc[inds, 'categ'] = 'clonal'
        print('Re-clustering subclones (Step 2)')
        df = cluster_subclones(df, sample, output_dir=os.path.dirname(output_file))
        df = clon_sub_distance(df)

    inds = df[(df['dist_clon'] > distance_scale * df['dist_sub']) & (~df['categ'].str.contains('subclonal')) & (df['dist_clon'] > 0) & (df['vaf'] < 1 / df['total_cn'])].index
    if len(inds) > 0:
        df.loc[inds, 'categ'] = 'subclonal1'
        print('Re-clustering subclones (Step 3)')
        df = cluster_subclones(df, sample, output_dir=os.path.dirname(output_file))
        df = clon_sub_distance(df)

    if df['categ'].str.contains('subclonal').sum() > n_sub_thresh:
        print('Final subclone clustering (Step 4)')
        df = cluster_subclones(df, sample, output_dir=os.path.dirname(output_file))

    df.to_csv(output_file, sep='\t', index=False)
    return 0

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Revise script for Tau workflow")
    parser.add_argument("sample", help="Sample name")
    parser.add_argument("input_file", help="Path to the input file")
    parser.add_argument("output_file", help="Path to the output file")
    args = parser.parse_args()
    revise(args.sample, args.input_file, args.output_file)
